ql/src/java/org/apache/hadoop/hive/ql/optimizer/DynamicPartitionPruningOptimization.java

```diff
@@ -678,38 +678,34 @@ private boolean generateSemiJoinOperatorPlan(DynamicListContext ctx, ParseContex
     ArrayList<ColumnInfo> groupbyColInfos = new ArrayList<ColumnInfo>();
     groupbyColInfos.add(new ColumnInfo(gbOutputNames.get(0), key.getTypeInfo(), "", false));
     groupbyColInfos.add(new ColumnInfo(gbOutputNames.get(1), key.getTypeInfo(), "", false));
-    groupbyColInfos.add(new ColumnInfo(gbOutputNames.get(2), key.getTypeInfo(), "", false));
+    groupbyColInfos.add(new ColumnInfo(gbOutputNames.get(2), TypeInfoFactory.binaryTypeInfo, "", false));
 
     GroupByOperator groupByOp = (GroupByOperator)OperatorFactory.getAndMakeChild(
             groupBy, new RowSchema(groupbyColInfos), selectOp);
 
     groupByOp.setColumnExprMap(new HashMap<String, ExprNodeDesc>());
 
     // Get the column names of the aggregations for reduce sink
-    int colPos = 0;
     ArrayList<ExprNodeDesc> rsValueCols = new ArrayList<ExprNodeDesc>();
     Map<String, ExprNodeDesc> columnExprMap = new HashMap<String, ExprNodeDesc>();
-    for (int i = 0; i < aggs.size() - 1; i++) {
-      ExprNodeColumnDesc colExpr = new ExprNodeColumnDesc(key.getTypeInfo(),
-              gbOutputNames.get(colPos), "", false);
+    ArrayList<ColumnInfo> rsColInfos = new ArrayList<>();
+    for (int colPos = 0; colPos < aggs.size(); colPos++) {
+      TypeInfo typInfo = groupbyColInfos.get(colPos).getType();
+      ExprNodeColumnDesc colExpr = new ExprNodeColumnDesc(typInfo, gbOutputNames.get(colPos), "", false);
       rsValueCols.add(colExpr);
-      columnExprMap.put(gbOutputNames.get(colPos), colExpr);
-      colPos++;
-    }
+      columnExprMap.put(Utilities.ReduceField.VALUE + "." + gbOutputNames.get(colPos), colExpr);
 
-    // Bloom Filter uses binary
-    ExprNodeColumnDesc colExpr = new ExprNodeColumnDesc(TypeInfoFactory.binaryTypeInfo,
-        gbOutputNames.get(colPos), "", false);
-    rsValueCols.add(colExpr);
-    columnExprMap.put(gbOutputNames.get(colPos), colExpr);
-    colPos++;
+      ColumnInfo colInfo =
```

Comment:
why do we need this, can we use groupByOp.getSchema()?

Reply:
According to our records, groupByOp.getSchema() does not contains `Utilities.ReduceField.VALUE`, which is required by RS operator.

Reply:
sorry for the dumb question, but why do we need `ReduceField.VALUE` prefix in  RS operator? if that is required, how did it even work?

Reply:
(Please understand that I may be wrong as it has been a long time since I worked on this issue. I'll check this issue again and leave you a comment if any corrections are needed.)
RS operator forwards key-value pairs to child operator, and RS operators' columns are separated into KEY and VALUE by prepend `ReduceField.KEY` and `VALUE`.
According to our records, we encountered "cannot find filed _col0 from [0:key, 1:value]" issue when we set correct ColumnExprMap to GBY operator. We found that SEL operator generated by PEF tries to read `_col0`, but RS operator only provides `VALUE._col0`. So we prepend the column schema as PEF refers to it during SEL operator creation.
I can't remember why DPPOptimization works okay without `ReduceField.KEY/VALUE`. But our record says that we checked that other RS operator creator components also use `ReduceField.KEY/VALUE` when setting the schema of RS operator.

Reply:
I've removed the ReduceField.VALUE from generateSemiJoinOperatorPlan, but kept in createFinalRsForSemiJoinOp --> sharedwork_semi_2.q passed and produced. same result 
````
**generateSemiJoinOperatorPlan**
    for (int i = 0; i < aggs.size(); i++) {
      ExprNodeColumnDesc colExpr = new ExprNodeColumnDesc(groupbyColInfos.get(i).getType(), 
              gbOutputNames.get(i), "", false);
      rsValueCols.add(colExpr);
      columnExprMap.put(gbOutputNames.get(i), colExpr);
    }
**createFinalRsForSemiJoinOp**
    ArrayList<ColumnInfo> gbySchema = new ArrayList<>();
    Map<String, ExprNodeDesc> columnExprMap = new HashMap<>();
    for (int i = 0; i < gb.getSchema().getSignature().size(); i++) {
      ColumnInfo rsColInfo = gb.getSchema().getSignature().get(i);
      ExprNodeColumnDesc rsValExpr = new ExprNodeColumnDesc(rsColInfo.getType(), rsColInfo.getInternalName(), "", false);
      rsValueCols.add(rsValExpr);

      String outputColName = SemanticAnalyzer.getColumnInternalName(i);
      gbOutputNames.add(outputColName);
      
      String field = Utilities.ReduceField.VALUE.toString() + "." + outputColName;
      columnExprMap.put(field, rsValExpr);
      gbySchema.add(new ColumnInfo(field, rsColInfo.getType(), "", false));
    }
````
also that didn't require changes in ParallelEdgeFixer.colMappingInverseKeys, exprMap wasn't empty

Reply:
@ngsg, patch LGTM, however, could we please keep just the min required changes and do a minor refactor as in the above snippet?

Reply:
thank you @kgyrtkirk for chiming in! 
I don't have much expertise in that area. Small question: 
When in FinalRsForSemiJoinOp we don't prefix columns with  ReduceField.VALUE (columnExprMap & schema) then ParallelEdgeFixer kicks in and introduces a concentrator RS that expects [key, value] inputs. 
````
Caused by: java.lang.RuntimeException: cannot find field _col0 from [0:key, 1:value]
	at org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorUtils.getStandardStructFieldRef(ObjectInspectorUtils.java:550)
	at org.apache.hadoop.hive.serde2.objectinspector.StandardStructObjectInspector.getStructFieldRef(StandardStructObjectInspector.java:153)
	at org.apache.hadoop.hive.ql.exec.ExprNodeColumnEvaluator.initialize(ExprNodeColumnEvaluator.java:56)
	at org.apache.hadoop.hive.ql.exec.Operator.initEvaluators(Operator.java:1073)
	at org.apache.hadoop.hive.ql.exec.Operator.initEvaluatorsAndReturnStruct(Operator.java:1099)
	at org.apache.hadoop.hive.ql.exec.SelectOperator.initializeOp(SelectOperator.java:74)
	at org.apache.hadoop.hive.ql.exec.Operator.initialize(Operator.java:360)
	at org.apache.hadoop.hive.ql.exec.tez.ReduceRecordProcessor.init(ReduceRecordProcessor.java:191)
	at org.apache.hadoop.hive.ql.exec.tez.TezProcessor.initializeAndRunProcessor(TezProcessor.java:292)
````
Should we prefix the columns in FinalRsForSemiJoinOp or make some fixes in fixParallelEdge SEL inputs?

Reply:
@ngsg could you please check the comment from https://github.com/apache/hive/pull/4043#discussion_r1388649716, other than that it LGTM

Reply:
@deniskuzZ, I have checked your comment and my work, and I summarized my conclusion as follows:

1. about `ReduceField.VALUE`

I think we should prepend the name of RS operator's columns in colExprMap and schema because RS's child operators always access to input columns(output of RS) as `KEY.col` and `VALUE.col`.

RS operator's output rows are transported to next operator via shuffle, not by directly calling `Operator.forward()`. `ReduceRecordSource` reads shuffled KV pairs and calls the child operator's `Operator.process()` on behalf of `Operator.forward()`. If vectorization is disabled, it passes a `List<Object>` of length 2 as a row, and the corresponding ObjectInspector consists of `ReduceField.KEY` and `ReduceField.VALUE`. [1] If vectorization is enabled, it passes a single struct object as a row, and the corresponding ObjectInspector consists of `ReduceField.KEY + "." + fieldName` and `ReduceField.VALUE + "." + fieldName`. [2] In both cases, the name of columns come from RS's child operator start with either `ReduceField.KEY` or `ReduceField.VALUE`.

`colExprMap` maps an output column name to its expression [3], so the key of `colExprMap` of RS should be prefixed by `ReduceField.KEY` or `ReduceField.VALUE`. I could not find any information about `schema`, but it seems that `schema` also represents output column names. [4] So I think both `colExprMap`'s key and `schema` of RS should be prefixed by `ReduceField.KEY` or `ReduceField.VALUE`.

2. about your investigation

DPPOptimization creates 4 operators, GBY->RS->GBY->RS, and `sharedwork_semi_2.q` tests PEF by inverting one of the final RS operators that DPPOptimization created. PEF refers to RS's `colExprMap` when creating a SEL operator that performs inversion. That's why we fail this test due to `java.lang.RuntimeException: cannot find field _col0 from [0:key, 1:value]` if we do not prepend the key of final RS's `colExprMap` with `ReduceField.VALUE`. Unlike final RS operators, intermediate RS operators are not inverted during the test,  so prepending intermediate RS operator's column name does not affect to the result of the test.

3. about `ParallelEdgeFixer.colMappingInverseKeys()`

According to the comment of Operator.getColumnExprMap(), it returns only key columns for RS operators. [3] I'm not sure whether it is still valid or not, but I want to keep the added code as a kind of defensive programming. 

[1] https://github.com/apache/hive/blob/8a4f5ce7275842ff4f1cc917c7a2a48dde71bf4c/ql/src/java/org/apache/hadoop/hive/ql/exec/tez/ReduceRecordSource.java#L229-L231
[2] https://github.com/apache/hive/blob/8a4f5ce7275842ff4f1cc917c7a2a48dde71bf4c/ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java#L4333-L4341
[3]
https://github.com/apache/hive/blob/8a4f5ce7275842ff4f1cc917c7a2a48dde71bf4c/ql/src/java/org/apache/hadoop/hive/ql/exec/Operator.java#L991-L997
[4]
https://github.com/apache/hive/blob/8a4f5ce7275842ff4f1cc917c7a2a48dde71bf4c/ql/src/java/org/apache/hadoop/hive/ql/optimizer/CountDistinctRewriteProc.java#L443-L447

Reply:
hi @ngsg, thanks for the extended explanation!
1. That's clear, I've debugged the code as well, thanks for sharing.
2. Unless you have a test that emulates PEF applied to the intermediate RS operator, please restore the original code. Also, could you please apply code refactor done in the snipped above.
3. Ok

Reply:
@ngsg, there is still 1 leftover + no need for index in for-loop (added comment). Please take a look